enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,474 @@
+      "network": "10-5-132-0-25",
+      "datastore": "DC1_DS1",
+      "resourcepool": "/DC1/host/DC1_C1/Resources"
+    }
+  }
+}
+```
+
+#### Machine and MachineSet
+
+The control plane
+[Machines](https://github.com/openshift/installer/blob/b0b96468893db2240e82ba2aa0935679c8c49201/pkg/asset/machines/vsphere/machines.go#L19-L64)
+will need to be modified to create a Machine per `region` and `zone` pair.
+
+For each `region` and `zone` pair an additional `MachineSet` for
+the compute nodes will need to be created.
+
+This is to provide the proper configuration to provision a virtual machine in
+each `zone`.
+
+##### OVA import
+
+For each `zone` in `Zones` a ova import will need to occur.
+If there is only a single zone then a single import will be required.
+
+### User Stories
+
+- https://issues.redhat.com/browse/RFE-845
+- https://issues.redhat.com/browse/OCPPLAN-4927
+- https://issues.redhat.com/browse/SPLAT-320
+
+### API Extensions
+
+
+### Risks and Mitigations
+
+- The out-of-tree CCM is required for this work. It will need to be enabled at
+installation time.
+
+## Design Details
+
+The vSphere platform spec and configuration will change to include `vcenters`
+and `regions`. The `MachinePool` will also add an optional `zones` field.
+
+`regions` will contain:
+
+- datacenter
+- vcenter name
+
+`zones` will contain:
+
+- cluster
+- resourcepool
+- datastore
+- network (port group)
+
+which are the required parameters to provision virtual machines across
+datacenters and clusters within a single vcenter.
+
+The vSphere platform spec will be the default configuration for the cloud-config
+adding the additional datacenters from the machinepool.
+
+The master virtual machines will be provisioned with terraform per datacenter
+and cluster. In the case of workers multiple machinesets will need to be
+configured - one per datacenter/cluster pair.
+
+```yaml
+apiVersion: v1
+baseDomain: example.com
+controlPlane:
+  name: master
+  platform:
+    vsphere:
+      zones:
+      - region-dc1-zone-c1
+      - region-dc1-zone-c2
+      - region-dc1-zone-c3
+      cpus: 8
+      coresPerSocket: 2
+      memoryMB: 24576
+      osDisk:
+        diskSizeGB: 512
+  replicas: 3
+compute:
+- name: worker
+  platform:
+    vsphere:
+      zones:
+      - region-dc1-zone-c1
+      - region-dc1-zone-c2
+      - region-dc1-zone-c3
+      cpus: 8
+      coresPerSocket: 2
+      memoryMB: 24576
+      osDisk:
+        diskSizeGB: 512
+  replicas: 3
+metadata:
+  name: test-cluster
+platform:
+  vSphere:
+    vCenter: your.vcenter.example.com
+    username: username
+    password: password
+    datacenter: datacenter
+    defaultDatastore: datastore
+    network: portgroup
+    vcenters:
+    - server: your.first-vcenter.example.com
+      user: username
+      password: password
+      regions:
+      - name: region-dc1
+        datacenter: dc1
+        vcenter: your.first-vcenter.example.com
+        zones:
+        - name: region-dc1-zone-c1
```

Comment:
Are there any caveats on zone names, for example, do they need need to be unique across all regions for example?

Reply:
Yes, zones need to be unique across all regions.

Reply:
Can we make sure that's noted somewhere and if possible, that we have validation of that during cluster bootstrapping. It makes me think that we can probably just have an `availabilityZone` field on the Machine API spec which then looks across all regions to find the appropriately named zone

Reply:
Was this note ever added?